### PR TITLE
🤖  Quarot

### PR DESCRIPTION
**Pull request by trajepl**
## Describe your changes

1. Integration QuaRot into olive, which is a new Quantization scheme based on Rotations, which is able to quantize LLMs end-to-end. 
2. Add phi3-quarot example

Release Note: New pass QuaRot to quantize transformer to improve performance and reduce memory footprint.

Example:
`w_gptq` needs flash-atten which only supports Ampere GPUs or newer.

Test on A100:
image

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [x] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`
- [x] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [x] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

### CODE REVIEW

File: olive/passes/pytorch/quarot.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+    @staticmethod
+    def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        return {
+            "input_model_dtype": PassConfigParam(
+                type_=QuaRot.ModelDtype,
+                default_value=QuaRot.ModelDtype.FP16,
+                description="Input model's data type.",
+            ),
+            "calibration_data_config": PassConfigParam(
```
Comment (jambayk) [inline olive/passes/pytorch/quarot.py:61]: Hitesh and I had a related discussion before. Would it be possible to add the functionality for sampling directly to our dataloader component so that this and slicegpt pass don't have to expose `calibration_nsamples` and `calibration_batch_size`? 
wrt the `"labels"` as inputs, we could just wrap the original dataloader and override its iteration logic to return a single dict.

Comment (trajepl) [reply olive/passes/pytorch/quarot.py:61]: Sounds good. How about implement it in follow-up PR?

Comment (jambayk) [reply olive/passes/pytorch/quarot.py:61]: Sure! Follow up is good too.

**Review by jambayk — APPROVED** [review]
